1. Setup & Load

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('../data/shipments.csv', parse_dates=[
    'booking_date', 'pickup_date', 'delivery_date',
    'promised_delivery_date', 'actual_delivery_date'
])

print(df.shape)
df.head()

(5015, 15)


,shipment_id,booking_date,pickup_date,delivery_date,origin_city,destination_city,region,mode,carrier_id,customer_id,promised_delivery_date,actual_delivery_date,freight_cost,distance_km,status
0,SHP003604,2026-04-18,2026-04-21,2026-04-27,Chandigarh,Guwahati,North,PTL,CARR_07,CUST_085,2026-04-27,2026-04-29,72921.16,894,Delivered
1,SHP003362,2026-06-01,2026-06-01,2026-06-04,Kolkata,Raipur,East,FTL,CARR_03,CUST_108,2026-06-04,2026-06-03,2451.90,91,Delivered
2,SHP000796,2026-01-17,2026-01-20,2026-01-23,Surat,Jaipur,West,FTL,CARR_05,CUST_108,2026-01-23,NaT,10320.66,395,In-Transit
3,SHP000291,2026-04-25,2026-04-25,2026-05-01,Nagpur,Bhubaneswar,Central,FTL,CARR_09,CUST_063,2026-05-01,2026-05-02,30242.52,1223,Delayed
4,SHP003739,2026-01-29,2026-01-30,2026-02-03,Ahmedabad,Hyderabad,West,LTL,CARR_04,CUST_064,2026-02-03,2026-02-01,18434.26,1664,Delivered


## 2. Data Profiling & Quality Checks

In [2]:
df.info()
df.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5015 entries, 0 to 5014
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   shipment_id             5015 non-null   object        
 1   booking_date            4944 non-null   datetime64[ns]
 2   pickup_date             4927 non-null   datetime64[ns]
 3   delivery_date           5015 non-null   datetime64[ns]
 4   origin_city             5015 non-null   object        
 5   destination_city        5015 non-null   object        
 6   region                  5015 non-null   object        
 7   mode                    5015 non-null   object        
 8   carrier_id              5015 non-null   object        
 9   customer_id             5015 non-null   object        
 10  promised_delivery_date  5015 non-null   datetime64[ns]
 11  actual_delivery_date    3527 non-null   datetime64[ns]
 12  freight_cost            5015 non-null   float64 

,shipment_id,booking_date,pickup_date,delivery_date,origin_city,destination_city,region,mode,carrier_id,customer_id,promised_delivery_date,actual_delivery_date,freight_cost,distance_km,status
count,5015,4944,4927,5015,5015,5015,5015,5015,5015,5015,5015,3527,5015.000000,5015.000000,5015
unique,5000,NaN,NaN,NaN,20,20,5,3,15,120,NaN,NaN,NaN,NaN,4
top,SHP001121,NaN,NaN,NaN,Surat,Lucknow,West,FTL,CARR_11,CUST_085,NaN,NaN,NaN,NaN,Delivered
freq,2,NaN,NaN,NaN,279,285,1052,2043,364,60,NaN,NaN,NaN,NaN,3616
mean,NaN,2026-04-01 13:28:15.145630976,2026-04-02 20:19:37.836411648,2026-04-07 23:10:36.729810688,NaN,NaN,NaN,NaN,NaN,NaN,2026-04-07 23:10:36.729810688,2026-04-07 19:52:34.975900160,33729.236782,1280.035494,NaN
min,NaN,2026-01-01 00:00:00,2026-01-01 00:00:00,2026-01-04 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,2026-01-04 00:00:00,2026-01-03 00:00:00,412.310000,50.000000,NaN
25%,NaN,2026-02-15 18:00:00,2026-02-17 00:00:00,2026-02-22 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,2026-02-22 00:00:00,2026-02-22 00:00:00,9511.410000,655.500000,NaN
50%,NaN,2026-04-02 00:00:00,2026-04-03 00:00:00,2026-04-08 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,2026-04-08 00:00:00,2026-04-07 00:00:00,18014.500000,1280.000000,NaN
75%,NaN,2026-05-18 00:00:00,2026-05-19 00:00:00,2026-05-24 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,2026-05-24 00:00:00,2026-05-23 00:00:00,32422.895000,1901.000000,NaN
max,NaN,2026-06-30 00:00:00,2026-07-03 00:00:00,2026-07-11 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,2026-07-11 00:00:00,2026-07-13 00:00:00,788479.360000,2499.000000,NaN


In [4]:
dupes = df[df.duplicated(subset='shipment_id', keep=False)].sort_values('shipment_id')
print(dupes.shape)
dupes.head(20)

(30, 15)


,shipment_id,booking_date,pickup_date,delivery_date,origin_city,destination_city,region,mode,carrier_id,customer_id,promised_delivery_date,actual_delivery_date,freight_cost,distance_km,status
430,SHP000280,2026-05-30,NaT,2026-06-09,Delhi,Indore,North,LTL,CARR_09,CUST_118,2026-06-09,2026-06-07,13019.33,1182,Delayed
2636,SHP000280,2026-05-30,NaT,2026-06-09,Delhi,Indore,North,LTL,CARR_09,CUST_118,2026-06-09,2026-06-07,13019.33,1182,Delayed
4572,SHP000445,2026-03-24,2026-03-27,2026-04-03,Hyderabad,Raipur,South,FTL,CARR_08,CUST_018,2026-04-03,NaT,43032.71,1620,In-Transit
3216,SHP000445,2026-03-24,2026-03-27,2026-04-03,Hyderabad,Raipur,South,FTL,CARR_08,CUST_018,2026-04-03,NaT,43032.71,1620,In-Transit
302,SHP000517,2026-01-03,2026-01-06,2026-01-10,Mumbai,Patna,West,LTL,CARR_03,CUST_040,2026-01-10,2026-01-12,11392.79,973,Delivered
4723,SHP000517,2026-01-03,2026-01-06,2026-01-10,Mumbai,Patna,West,LTL,CARR_03,CUST_040,2026-01-10,2026-01-12,11392.79,973,Delivered
277,SHP000611,2026-05-09,2026-05-10,2026-05-16,Bengaluru,Guwahati,South,PTL,CARR_09,CUST_013,2026-05-16,NaT,12830.67,1615,Cancelled
3410,SHP000611,2026-05-09,2026-05-10,2026-05-16,Bengaluru,Guwahati,South,PTL,CARR_09,CUST_013,2026-05-16,NaT,12830.67,1615,Cancelled
1917,SHP000625,2026-05-19,2026-05-22,2026-05-26,Kochi,Ahmedabad,South,FTL,CARR_10,CUST_021,2026-05-26,2026-05-29,47876.69,1841,Delivered
2143,SHP000625,2026-05-19,2026-05-22,2026-05-26,Kochi,Ahmedabad,South,FTL,CARR_10,CUST_021,2026-05-26,2026-05-29,47876.69,1841,Delivered


In [8]:
exact_dupes = df[df.duplicated(keep=False)]
print("Exact full-row duplicates:", exact_dupes.shape[0])
print("Same ID but different data:", dupes.shape[0] - exact_dupes.shape[0])

df.isnull().sum()

# df[df['actual_delivery_date'].isnull()]['status'].value_counts()

Exact full-row duplicates: 30
Same ID but different data: 0


shipment_id                  0
booking_date                71
pickup_date                 88
delivery_date                0
origin_city                  0
destination_city             0
region                       0
mode                         0
carrier_id                   0
customer_id                  0
promised_delivery_date       0
actual_delivery_date      1488
freight_cost                 0
distance_km                  0
status                       0
dtype: int64

In [9]:
df[df['actual_delivery_date'].isnull()]['status'].value_counts()

status
Delivered     588
In-Transit    502
Cancelled     302
Delayed        96
Name: count, dtype: int64

**Finding:** Of 1,488 missing `actual_delivery_date` rows, 502 are In-Transit and 302 are Cancelled — both expected. But 588 rows marked **Delivered** and 96 marked **Delayed** have no actual_delivery_date — a genuine status/data inconsistency affecting ~13.6% of the dataset. These rows are excluded from on-time% and delay_days calculations rather than imputed.

In [10]:
df[df['booking_date'].isnull()].head(10)
df[df['pickup_date'].isnull()].head(10)

,shipment_id,booking_date,pickup_date,delivery_date,origin_city,destination_city,region,mode,carrier_id,customer_id,promised_delivery_date,actual_delivery_date,freight_cost,distance_km,status
56,SHP001316,2026-01-03,NaT,2026-01-07,Raipur,Bhubaneswar,Central,LTL,CARR_03,CUST_092,2026-01-07,2026-01-06,11151.07,866,Delivered
75,SHP003329,2026-06-24,NaT,2026-07-02,Bhopal,Indore,Central,LTL,CARR_14,CUST_117,2026-07-02,2026-07-02,5758.35,532,Delayed
133,SHP003752,2026-01-26,NaT,2026-01-30,Kolkata,Chennai,East,LTL,CARR_10,CUST_007,2026-01-30,2026-01-23,8631.21,661,Delivered
138,SHP002877,2026-06-18,NaT,2026-06-23,Ahmedabad,Lucknow,West,FTL,CARR_06,CUST_089,2026-06-23,2026-06-28,32300.15,1392,Delivered
148,SHP004278,2026-05-20,NaT,2026-05-25,Bhubaneswar,Hyderabad,East,LTL,CARR_14,CUST_051,2026-05-25,2026-05-30,17565.38,1623,Delivered
157,SHP003130,2026-06-03,NaT,2026-06-08,Ahmedabad,Hyderabad,West,PTL,CARR_11,CUST_083,2026-06-08,2026-06-08,19615.08,2303,Delivered
163,SHP000254,2026-06-16,NaT,2026-06-23,Jaipur,Raipur,North,LTL,CARR_05,CUST_115,2026-06-23,2026-06-23,5268.45,387,Delivered
164,SHP004743,2026-04-14,NaT,2026-04-23,Jaipur,Bengaluru,North,LTL,CARR_14,CUST_055,2026-04-23,2026-04-24,26047.86,2225,Delivered
193,SHP001390,2026-04-18,NaT,2026-04-23,Jaipur,Ahmedabad,North,LTL,CARR_15,CUST_088,2026-04-23,2026-04-26,6743.85,581,Delivered
206,SHP002885,NaT,NaT,2026-02-19,Ahmedabad,Raipur,West,PTL,CARR_09,CUST_002,2026-02-19,2026-02-23,10625.97,1361,Delayed


**Finding:** `booking_date` (71 rows, 1.4%) and `pickup_date` (88 rows, 1.8%) missing values show no discernible pattern — treated as random gaps. Excluded from `pickup_lag_days` calculations where null, retained elsewhere.

In [14]:
# Delivered but no actual date - the real issue
delivered_no_date = df[(df['status']=='Delivered') & (df['actual_delivery_date'].isnull())]
print(delivered_no_date.shape)

# Delayed but no actual date - also odd
delayed_no_date = df[(df['status']=='Delayed') & (df['actual_delivery_date'].isnull())]
print(delayed_no_date.shape)

# Sanity check the reverse - any In-Transit WITH an actual date? (also a red flag if so)
intransit_with_date = df[(df['status']=='In-Transit') & (df['actual_delivery_date'].notnull())]
print(intransit_with_date.shape)

(588, 15)
(96, 15)
(0, 15)


**Finding:** Confirms 588 Delivered and 96 Delayed shipments are missing actual_delivery_date (a status/data pipeline inconsistency), while In-Transit rows correctly never have an actual_delivery_date recorded. Recommend flagging this to the data/ops team as a pipeline integrity issue.

In [17]:
from ydata_profiling import ProfileReport

profile = ProfileReport(df, title="Shipments Data Profiling Report", explorative=True)
profile.to_file("../docs/shipments_profile_report.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]


100%|██████████| 15/15 [00:00<00:00, 83.32it/s][A


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [44]:
import webbrowser
webbrowser.open('../docs/shipments_profile_report.html')

True

3. Data Cleaning and derived  features

In [20]:
df_clean = df.drop_duplicates(keep='first').copy()

# Only compute on_time where actual_delivery_date exists AND status is a completed one
df_clean['on_time'] = np.where(
    df_clean['actual_delivery_date'].notnull(),
    df_clean['actual_delivery_date'] <= df_clean['promised_delivery_date'],
    np.nan
)

df_clean['delay_days'] = (df_clean['actual_delivery_date'] - df_clean['promised_delivery_date']).dt.days
df_clean['transit_time_days'] = (df_clean['actual_delivery_date'] - df_clean['pickup_date']).dt.days
df_clean['pickup_lag_days'] = (df_clean['pickup_date'] - df_clean['booking_date']).dt.days
df_clean['cost_per_km'] = df_clean['freight_cost'] / df_clean['distance_km']

df_clean.to_csv('../data/shipments_clean.csv', index=False)
print(df_clean.shape)

(5000, 20)


**Data quality summary (Q4):**- 30 exact full-row duplicates dropped (0.6% of data)- booking_date / pickup_date: minor random gaps (~1.5-1.8%), excluded from lag calcs where null- Status/actual_delivery_date mismatch: 684 rows (13.6%) — Delivered/Delayed shipments missing actual_delivery_date — excluded from on-time% and delay_days rather than imputed, to avoid silently biasing headline metrics- Severe right-skew in freight_cost (skew ≈ 5.9) — used log-transform and median alongside mean throughout to avoid misleading averages- CARR_07 cost anomaly (see Q2 below) — likely a data/unit error, not genuine pricing

## Q1: Region vs Carrier Analysis (which region has worst OTP, and why?)

In [21]:
# On-time % by region (only on non-null on_time rows)
region_otp = df_clean.dropna(subset=['on_time']).groupby('region')['on_time'].mean().sort_values()
print(region_otp)

# Decompose: region x carrier
region_carrier_otp = df_clean.dropna(subset=['on_time']).groupby(['region','carrier_id'])['on_time'].agg(['mean','count'])
print(region_carrier_otp.sort_values('mean').head(15))

region
Central    0.497674
North      0.507822
East       0.512575
West       0.520185
South      0.528000
Name: on_time, dtype: float64
                        mean  count
region  carrier_id                 
South   CARR_02     0.300000     10
        CARR_05     0.333333      3
        CARR_14     0.333333      6
        CARR_11     0.333333      9
        CARR_10     0.363636     11
Central CARR_08     0.388889     54
West    CARR_02     0.391304     69
        CARR_13     0.392857     56
North   CARR_03     0.400000     40
East    CARR_07     0.400000     70
Central CARR_02     0.414286     70
South   CARR_07     0.428571      7
North   CARR_13     0.431034     58
        CARR_02     0.431373     51
East    CARR_06     0.436364     55


In [22]:
region_carrier_otp = df_clean.dropna(subset=['on_time']).groupby(['region','carrier_id'])['on_time'].agg(['mean','count'])
reliable = region_carrier_otp[region_carrier_otp['count'] >= 20].sort_values('mean')
print(reliable.head(15))

                        mean  count
region  carrier_id                 
Central CARR_08     0.388889     54
West    CARR_02     0.391304     69
        CARR_13     0.392857     56
East    CARR_07     0.400000     70
North   CARR_03     0.400000     40
Central CARR_02     0.414286     70
North   CARR_13     0.431034     58
        CARR_02     0.431373     51
East    CARR_06     0.436364     55
North   CARR_08     0.448276     58
Central CARR_07     0.450000     40
East    CARR_05     0.460317     63
Central CARR_10     0.462963     54
East    CARR_12     0.463415     41
        CARR_02     0.466667     45


In [23]:
carrier_otp = df_clean.dropna(subset=['on_time']).groupby('carrier_id')['on_time'].agg(['mean','count']).sort_values('mean')
print(carrier_otp)

                mean  count
carrier_id                 
CARR_02     0.416327    245
CARR_07     0.458515    229
CARR_13     0.461538    234
CARR_08     0.471861    231
CARR_06     0.490566    212
CARR_05     0.510460    239
CARR_04     0.513514    222
CARR_03     0.513514    222
CARR_01     0.514056    249
CARR_10     0.537815    238
CARR_12     0.545894    207
CARR_09     0.546875    256
CARR_15     0.550607    247
CARR_14     0.555556    225
CARR_11     0.561069    262


In [26]:
carrier_region_mix = pd.crosstab(df_clean['carrier_id'], df_clean['region'], normalize='index')
print(carrier_region_mix.loc[['CARR_02','CARR_07','CARR_13','CARR_08']].round(2))

region      Central  East  North  South  West
carrier_id                                   
CARR_02        0.23  0.16   0.18   0.22  0.21
CARR_07        0.13  0.24   0.21   0.22  0.19
CARR_13        0.22  0.20   0.22   0.18  0.19
CARR_08        0.20  0.18   0.21   0.20  0.21


**Q1 Answer:** Regional OTP spread is small (49.8%–54.8%, a 3-point range) — no region stands out as meaningfully worse. Carrier OTP spread is far wider (41.6%–56.1%, a 14.5-point range), nearly 5x the regional spread. The weakest carriers (CARR_02, CARR_07, CARR_13, CARR_08) operate at similarly low rates across *all* five regions (13–24% share each, roughly even) rather than being concentrated in one region — confirmed by removing them from the region calculation, which barely changed regional gaps (3.0 → 4.3 points). **Conclusion: poor OTP is a carrier-driven, company-wide problem, not a regional one.** Recommend a carrier performance review rather than a regional intervention.

## Q2: Freight Cost vs Distance (relationship, and which carrier deviates?)

In [27]:
import statsmodels.api as sm

X = sm.add_constant(df_clean['distance_km'])
y = df_clean['freight_cost']
model = sm.OLS(y, X, missing='drop').fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:           freight_cost   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.087
Method:                 Least Squares   F-statistic:                     478.4
Date:                Tue, 28 Jul 2026   Prob (F-statistic):          2.33e-101
Time:                        13:31:16   Log-Likelihood:                -62317.
No. Observations:                5000   AIC:                         1.246e+05
Df Residuals:                    4998   BIC:                         1.247e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1078.0083   1821.578     -0.592      

**Finding:** Distance alone is a weak predictor of freight cost (R² = 0.087).

In [28]:
df_clean.groupby('mode')['freight_cost'].describe()
df_clean.groupby('mode')['cost_per_km'].describe()

,count,mean,std,min,25%,50%,75%,max
mode,,,,,,,,
FTL,2033.0,39.331341,55.704469,21.250794,23.215708,25.335782,27.194191,335.781630
LTL,1982.0,19.298762,27.111491,10.203923,11.178144,12.145412,13.080899,158.563827
PTL,985.0,13.332979,19.125842,6.803841,7.402143,8.026900,8.674733,107.049312


In [29]:
import statsmodels.formula.api as smf

model2 = smf.ols('freight_cost ~ distance_km + C(mode)', data=df_clean).fit()
print(model2.summary())

                            OLS Regression Results                            
Dep. Variable:           freight_cost   R-squared:                       0.138
Model:                            OLS   Adj. R-squared:                  0.137
Method:                 Least Squares   F-statistic:                     265.6
Date:                Tue, 28 Jul 2026   Prob (F-statistic):          6.47e-160
Time:                        13:34:14   Log-Likelihood:                -62176.
No. Observations:                5000   AIC:                         1.244e+05
Df Residuals:                    4996   BIC:                         1.244e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept       1.605e+04   2051.948      7.

**Finding:** Adding transport mode improves the model (R² = 0.138) — cost is jointly driven by distance and mode.

In [30]:
df_clean['log_cost'] = np.log(df_clean['freight_cost'])
model3 = smf.ols('log_cost ~ distance_km + C(mode)', data=df_clean).fit()
print(model3.summary())

                            OLS Regression Results                            
Dep. Variable:               log_cost   R-squared:                       0.637
Model:                            OLS   Adj. R-squared:                  0.636
Method:                 Least Squares   F-statistic:                     2917.
Date:                Wed, 29 Jul 2026   Prob (F-statistic):               0.00
Time:                        16:13:50   Log-Likelihood:                -5019.7
No. Observations:                5000   AIC:                         1.005e+04
Df Residuals:                    4996   BIC:                         1.007e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          8.8964      0.022    399.

**Finding:** A log-linear model (log(freight_cost) ~ distance + mode) fits far better, R² = 0.637 — consistent with the severe right-skew (skew ≈ 5.9) in raw freight_cost. This is the model used for carrier residual comparisons below.

In [31]:
df_clean['predicted_cost'] = model2.predict(df_clean)
df_clean['cost_residual'] = df_clean['freight_cost'] - df_clean['predicted_cost']

carrier_residuals = df_clean.groupby('carrier_id')['cost_residual'].agg(['mean','count']).sort_values('mean')
print(carrier_residuals)

                     mean  count
carrier_id                      
CARR_04     -13612.010501    319
CARR_13     -13551.360128    309
CARR_01     -13171.798464    352
CARR_10     -13092.099124    330
CARR_15     -13075.740544    361
CARR_06     -12885.956698    320
CARR_05     -12875.335797    329
CARR_09     -12786.660068    344
CARR_12     -12696.729707    312
CARR_14     -12613.319491    323
CARR_11     -12506.246508    362
CARR_02     -12466.696352    360
CARR_03     -11477.149086    321
CARR_08     -11245.767055    316
CARR_07     173255.828808    342


In [32]:
carrier_residuals_full = df_clean.groupby('carrier_id')['cost_residual'].agg(['mean','median','std','count']).sort_values('mean')
print(carrier_residuals_full)

                     mean         median            std  count
carrier_id                                                    
CARR_04     -13612.010501  -16410.060646   10029.144153    319
CARR_13     -13551.360128  -16673.990543   10923.702629    309
CARR_01     -13171.798464  -16485.909896   10453.335784    352
CARR_10     -13092.099124  -16513.786353   11025.862113    330
CARR_15     -13075.740544  -16135.337238   10798.159426    361
CARR_06     -12885.956698  -16414.187350   11028.466560    320
CARR_05     -12875.335797  -15897.011192   10266.080131    329
CARR_09     -12786.660068  -16178.239137   11189.882694    344
CARR_12     -12696.729707  -16418.590003   10990.392954    312
CARR_14     -12613.319491  -16204.187680   11239.062839    323
CARR_11     -12506.246508  -16188.452650   11440.397761    362
CARR_02     -12466.696352  -15848.096108   10734.006789    360
CARR_03     -11477.149086  -15843.331023   11284.442580    321
CARR_08     -11245.767055  -15593.134421   10945.978604

In [33]:
carr07 = df_clean[df_clean['carrier_id']=='CARR_07'].sort_values('cost_residual', ascending=False)
print(carr07[['shipment_id','mode','distance_km','freight_cost','predicted_cost','cost_residual']].head(15))

     shipment_id mode  distance_km  freight_cost  predicted_cost  \
2257   SHP004431  FTL         2404     788479.36    81889.324826   
510    SHP004046  FTL         2358     740721.22    80629.474945   
4561   SHP002053  FTL         2470     697103.14    83696.935525   
4536   SHP002064  FTL         2312     684098.53    79369.625064   
674    SHP000151  FTL         2489     668944.83    84217.308301   
2290   SHP001474  FTL         2400     646267.43    81779.772662   
800    SHP003688  FTL         2149     627984.49    74905.374399   
3229   SHP000676  FTL         2213     626091.56    76658.209016   
2949   SHP004797  FTL         2003     597691.88    70906.720429   
3736   SHP004315  FTL         2436     598462.46    82765.742134   
520    SHP003505  FTL         2154     590687.91    75042.314604   
1365   SHP003707  FTL         2333     594998.89    79944.773923   
4751   SHP001797  FTL         2098     583479.76    73508.584314   
552    SHP002687  FTL         2341     579175.36

In [34]:
print((carr07['cost_residual'] > 10000).sum(), "out of", len(carr07))

337 out of 342


In [35]:
df_clean['predicted_log_cost'] = model3.predict(df_clean)
df_clean['log_cost_residual'] = df_clean['log_cost'] - df_clean['predicted_log_cost']

carrier_log_residuals = df_clean.groupby('carrier_id')['log_cost_residual'].agg(['mean','median','count']).sort_values('mean')
print(carrier_log_residuals)

                mean    median  count
carrier_id                           
CARR_13    -0.184098 -0.110572    309
CARR_10    -0.182356 -0.082829    330
CARR_04    -0.180256 -0.092686    319
CARR_12    -0.165116 -0.087792    312
CARR_09    -0.159867 -0.072681    344
CARR_14    -0.158899 -0.083786    323
CARR_15    -0.154948 -0.043931    361
CARR_06    -0.152444 -0.083549    320
CARR_11    -0.151437 -0.083284    362
CARR_01    -0.142926 -0.077102    352
CARR_08    -0.140459 -0.051458    316
CARR_03    -0.138682 -0.062268    321
CARR_02    -0.136484 -0.059771    360
CARR_05    -0.112945 -0.026934    329
CARR_07     2.097789  2.175598    342


In [ ]:
# Is this only in FTL, or across all modes for CARR_07?
carr07_by_mode = df_clean[df_clean['carrier_id']=='CARR_07'].groupby('mode')['freight_cost'].describe()
print(carr07_by_mode)

# Compare to other carriers' FTL costs for similar distance
ftl_others = df_clean[(df_clean['mode']=='FTL') & (df_clean['carrier_id']!='CARR_07')]
print(ftl_others['cost_per_km'].describe())

carr07_ftl = df_clean[(df_clean['mode']=='FTL') & (df_clean['carrier_id']=='CARR_07')]
print(carr07_ftl['cost_per_km'].describe())

**Q2 Answer:** All carriers except one cluster tightly (log-residual range: -18% to -11% vs the model's expected cost) — consistent, unremarkable pricing. **CARR_07 is a severe outlier**: costs average ~700–780% above expected, affecting 337 of 342 shipments (98.5%) — a systemic pattern, not a few outliers. Directly comparing cost-per-km confirms this: CARR_07's FTL cost-per-km averages ~249 vs ~25 for every other carrier — almost exactly a 10x factor, consistent across the full min/max range. A clean ~10x scaling strongly suggests a **systematic data/unit error** in CARR_07's cost records rather than genuine pricing — no carrier could commercially sustain 10x market rates. Recommend the data/ops team verify CARR_07's cost entries at the source before using this figure in any carrier renegotiation decision.

## Q3: Customer Delay Concentration (which customers, and why?)

In [37]:
valid = df_clean.dropna(subset=['delay_days'])

customer_delay = valid.groupby('customer_id')['delay_days'].agg(['mean','median','count']).sort_values('mean', ascending=False)
print(customer_delay[customer_delay['count'] >= 5].head(15))

                 mean  median  count
customer_id                         
CUST_079     1.800000     2.0     25
CUST_119     1.656250     2.0     32
CUST_026     1.608696     2.0     23
CUST_051     1.590909     1.5     22
CUST_002     1.500000     2.0     34
CUST_109     1.391304     1.0     23
CUST_095     1.300000     1.0     30
CUST_014     1.259259     1.0     27
CUST_116     1.257143     2.0     35
CUST_054     1.241379     1.0     29
CUST_071     1.214286     1.5     28
CUST_028     1.208333     1.5     24
CUST_070     1.142857     1.5     28
CUST_010     1.115385     2.0     26
CUST_075     1.062500     1.0     32


In [39]:
top_customers = customer_delay[customer_delay['count'] >= 5].head(10).index.tolist()

top_cust_detail = valid[valid['customer_id'].isin(top_customers)].groupby(['customer_id','carrier_id']).size().unstack(fill_value=0)
print(top_cust_detail)

top_cust_region = valid[valid['customer_id'].isin(top_customers)].groupby(['customer_id','region']).size().unstack(fill_value=0)
print(top_cust_region)

carrier_id   CARR_01  CARR_02  CARR_03  CARR_04  CARR_05  CARR_06  CARR_07  \
customer_id                                                                  
CUST_002           3        2        3        1        2        3        0   
CUST_014           1        1        1        2        2        2        1   
CUST_026           1        2        0        3        2        0        3   
CUST_051           2        0        1        1        0        1        3   
CUST_054           3        5        1        3        2        0        0   
CUST_079           3        0        0        1        0        2        7   
CUST_095           0        4        2        2        1        3        3   
CUST_109           3        1        2        1        1        1        0   
CUST_116           2        4        5        2        4        1        1   
CUST_119           2        1        1        0        6        2        4   

carrier_id   CARR_08  CARR_09  CARR_10  CARR_11  CARR_12  CARR_

In [42]:
# Is delay simply proportional to volume? i.e., are "most delayed" customers just the ones who ship the most?
# valid.groupby('customer_id').size().corr(valid.groupby('customer_id')['delay_days'].mean())
cust_summary = valid.groupby('customer_id').agg(total_shipments=('delay_days','count'), avg_delay=('delay_days','mean'))
print(cust_summary['total_shipments'].corr(cust_summary['avg_delay']))

-0.1599944706591373


In [43]:
top_cust_mode = valid[valid['customer_id'].isin(top_customers)].groupby(['customer_id','mode']).size().unstack(fill_value=0)
print(top_cust_mode)

mode         FTL  LTL  PTL
customer_id               
CUST_002      16   12    6
CUST_014      14    8    5
CUST_026      11   10    2
CUST_051       9   10    3
CUST_054      13   11    5
CUST_079       8   12    5
CUST_095      11   12    7
CUST_109       9    9    5
CUST_116      16   17    2
CUST_119      16   10    6


**Q3 Answer:** The most delayed customers (e.g. CUST_079, CUST_119, CUST_026, averaging 1.6–1.8 days late) are **not** explained by a single dominant carrier — delay exposure is spread across 10+ carriers per customer, with only mild overrepresentation of CARR_07 for one customer (CUST_079: 28% of shipments vs ~7% baseline share). Delay is also not explained by shipment volume (correlation between total shipments and average delay = -0.16, negligible) or by transport mode mix (proportionally unremarkable). What stands out: these customers' shipments are heavily weighted toward Central, East, North, and West — the four weaker-performing regions from Q1 — with almost no shipments through South, the best-performing region. **Conclusion: delay concentration for these customers reflects their regional shipping mix, not a single carrier or customer-specific issue.** Recommend addressing the company-wide carrier reliability gap from Q1, which would benefit these customers as a side effect.

## Q5: Recommended Weekly Metric**Recommend: Weekly on-time delivery % broken out by carrier** (not a blended company-wide number).Since Q1 showed OTP is driven far more by carrier (14.5-point spread) than region (3-point spread), a blended weekly on-time % would mask carrier-specific problems the same way it currently masks anomalies like CARR_07. Tracking OTP by carrier weekly surfaces a carrier's decline in the same week it starts, well before a monthly aggregate would move enough to raise concern — and creates a direct audit trail for carrier contract reviews (tying into the Q2 cost-anomaly finding).**Secondary metric:** % of 'Delivered' shipments missing an actual_delivery_date — an early warning for the data pipeline issue found in Q4. A rising trend signals a tracking/system problem before it corrupts a month of delivery metrics.